flair/models/lemmatizer_model.py

```diff
@@ -0,0 +1,849 @@
+
+        self.alphabet_size = len(self.char_dictionary)
+
+        # decoder: linear layers to transform vectors to and from alphabet_size
+        self.character_embedding_for_decoding = nn.Embedding(self.alphabet_size, input_size)
+        if self.use_attention and token_embedding==None:
+            # when using attention we concatenate attention outcome and decoder hidden states
+            self.character_decoder = nn.Linear(2*self.hidden_size, self.alphabet_size)
+        else:
+            self.character_decoder = nn.Linear(self.hidden_size, self.alphabet_size)
+
+        # decoder RNN
+        self.input_size = input_size
+        self.num_layers = num_layers_in_rnn
+
+        self.decoding_rnn = nn.GRU(input_size=input_size, hidden_size=self.hidden_size, batch_first=True,
+                                   num_layers=num_layers_in_rnn)
+
+        # alternative encoder
+        self.encoding_rnn = nn.GRU(input_size=input_size, hidden_size=self.hidden_size, batch_first=True,
+                          num_layers=num_layers_in_rnn)
+
+        self.character_embedding_for_encoding = nn.Embedding(self.alphabet_size,input_size)
+
+        # loss and softmax
+        self.loss = nn.CrossEntropyLoss()
+        self.unreduced_loss = nn.CrossEntropyLoss(reduction='none') # for prediction
+        self.softmax = nn.Softmax(dim=2)
+
+        self.to(flair.device)
+
+    @property
+    def label_type(self):
+        return self._label_type
+
+    def words_to_char_indices(self, words: List[str], for_input=True, seq_length=None):
+        """
+        For a given list of labels (lemmas) this function creates index vectors that represent the characters of the single words in the sentence.
+        The "batch size" is given by the number of words in the list. Then each word is represented by sequence_length (maximum word length in the
+        list + 1)  many indices representing characters in self.char_dict.
+        We add one to the maximum_length since we add either a '1' at the beginning of the vectors (representing the start symbol '<S>') if for_input=True
+        or a '2' at the end of the vectors (representing the end symbol '<E>') if for_input = False.
+        One can manually set the vector length with the parameter seq_length, though the vector length is always at least maximum word length in the
+        list + 1
+        """
+        max_length= max(len(label) for label in words) + 1
+        if not seq_length:
+            # sequence length of each word is equal to max length of a word in the sentence plus one (start character <S>)
+            sequence_length = max_length
+        else:
+            sequence_length=max(seq_length,max_length)
+        # batch size is length of sentence
+        tensor = torch.ones(len(words), sequence_length, dtype=torch.long).to(flair.device)
+        for i in range(len(words)):
+            if for_input:
+                tensor[i][0] = 2  # start character <S>
```

Comment:
This is not robust to other dictionaries where the <S> token might have another index. Better never hard-code indices.

Reply:
Replaced by variables